File: runtime/src/main/java/com/huskerdev/nativekt/jvmci/Buffer.java
package com.huskerdev.nativekt.jvmci;

import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.util.Arrays;

public class Buffer {

    private ByteBuffer data = ByteBuffer.allocate(32).order(ByteOrder.nativeOrder());

    private void ensureSize(int length) {
        if (length >= data.limit()) {
            byte[] newBuf = Arrays.copyOf(data.array(), length * 4);
            ByteBuffer newData = ByteBuffer.wrap(newBuf);
            newData.order(data.order());
            newData.position(data.position());
            data = newData;
        }
    }

    public void emitByte(int b) {
        ensureSize(data.position() + 1);
        data.put((byte) (b & 0xFF));
    }

    public void emitInt(int b) {
        ensureSize(data.position() + 4);
        data.putInt(b);
    }

    public void emitLong(long b) {
        ensureSize(data.position() + 8);
        data.putLong(b);
    }

    public byte[] finish() {
        return Arrays.copyOf(data.array(), data.position());
    }
}

File: modules/runtime/src/jvmMain/kotlin/com/huskerdev/nativekt/jvm/jvmci/Buffer.kt
package com.huskerdev.nativekt.jvm.jvmci;

import java.nio.ByteBuffer
import java.nio.ByteOrder

class Buffer {
    private var data = ByteBuffer.allocate(32).order(ByteOrder.nativeOrder())

    private fun ensureSize(length: Int) {
        if (length >= data.limit()) {
            val newData = ByteBuffer.wrap(data.array().copyOf(data.limit() * 2))
            newData.order(data.order())
            newData.position(data.position())
            data = newData
        }
    }

    fun emitByte(b: Int) {
        ensureSize(data.position() + 1)
        data.put((b and 0xFF).toByte())
    }

    fun emitInt(b: Int) {
        ensureSize(data.position() + 4)
        data.putInt(b)
    }

    fun emitLong(b: Long) {
        ensureSize(data.position() + 8)
        data.putLong(b)
    }

    fun position() =
        data.position()

    fun finish() =
        data.array().copyOf(data.position())
}